### PR TITLE
Add validation of connectors HTTP methods

### PR DESCRIPTION
**Pull request by pubmodmatt**
* Verify that there is exactly one HTTP method on `@connect` `http`
* Modify unit tests to use the [glob feature](https://docs.rs/insta/1.39.0/insta/macro.glob.html) of insta to load schema files and store snapshot results in files
* Add unit tests for new validations

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[CNN-135]: https://apollographql.atlassian.net/browse/CNN-135?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by lennyburdette — APPROVED** [review]

**Review by dylan-apollo — CHANGES_REQUESTED** [review]
Thanks for the quick PR! Comments below on some suggestions—I think the only _required_ one is changing a code.

File: apollo-federation/src/sources/connect/models/validation.rs [hunk 1]
```diff
@@ -688,172 +734,171 @@ mod test_validate_source {
...
+        assert_eq!(errors.len(), 1);
+        assert_eq!(errors[0].code, Code::MultipleHttpMethods);
+        assert_snapshot!(errors[0].message, @r###"`@connect(http:)` on `Query.resources` cannot specify more than one HTTP method."###);
+    }
+
+    #[test]
+    fn missing_http_method_on_connect() {
+        let errors = validate!("test_data/missing_http_method_on_connect.graphql");
+        assert_eq!(errors.len(), 1);
+        assert_eq!(errors[0].code, Code::GraphQLError);
```
[the remark this answers is not in the document]
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs:893]: Will do! I went back and forth on this, so good to have a second opinion here!

File: apollo-federation/src/sources/connect/models/test_data/invalid_http_method_on_connect.graphql [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+@link(
+    url: "https://specs.apollo.dev/connect/v0.1"
+    import: ["@source", "@connect"]
+)
+@source(name: "v2", http: {baseURL: "http://127.0.0.1"})
+
+type Query {
+    resources: [String!]!
+    @connect(source: "v2", http: {
+        FOOBAR: "/resources"
```
[the remark this answers is not in the document]
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/test_data/invalid_http_method_on_connect.graphql]: I removed it, this was mostly a development artifact anyway.

File: apollo-federation/src/sources/connect/models/validation.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs]: Done.

File: apollo-federation/src/sources/connect/models/validation.rs [hunk 3]
```diff
@@ -311,9 +308,47 @@ fn validate_field(
         });
         return errors;
     };
+    let http_methods: Vec<_> = http_arg
+        .iter()
+        .filter(|(method, _)| ["GET", "POST", "PUT", "PATCH", "DELETE"].contains(&method.as_str()))
```
[the remark this answers is not in the document]
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs]: Done.

File: apollo-federation/src/sources/connect/models/validation.rs [hunk 4]
```diff
@@ -311,9 +308,47 @@ fn validate_field(
...
+            code: Code::MultipleHttpMethods,
+            message: format!(
+                "{coordinate} cannot specify more than one HTTP method.",
+                coordinate = connect_directive_http_coordinate(
+                    connect_directive_name,
+                    object_name,
+                    &field.name
+                ),
+            ),
+            locations: Location::from_node(connect_directive.location(), source_map)
```
[the remark this answers is not in the document]
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs]: Done.

File: apollo-federation/src/sources/connect/models/validation.rs [hunk 5]
```diff
@@ -311,9 +308,47 @@ fn validate_field(
...
+                .into_iter()
+                .collect(),
+        });
+    }
+    let Some((http_method, url)) = http_methods.first() else {
+        return errors;
+    };
```
[the remark this answers is not in the document]
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs]: Done.

File: licenses.html [hunk 6]
```diff
@@ -44,7 +44,7 @@ <h1>Third Party Licenses</h1>
     
         <h2>Overview of licenses:</h2>
         <ul class="licenses-overview">
-            <li><a href="#Apache-2.0">Apache License 2.0</a> (482)</li>
+            <li><a href="#Apache-2.0">Apache License 2.0</a> (485)</li>
```
Comment (dylan-apollo) [inline licenses.html:47]: I assume this happened automatically from some of our tooling? Related to adding the glob feature maybe? Probably want to double check with someone who knows better (maybe @lennyburdette) that there's no issue with this.

File: apollo-federation/src/sources/connect/models/validation.rs [hunk 7]
```diff
@@ -637,6 +703,7 @@ impl Location {
 }
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[non_exhaustive]
```
Comment (dylan-apollo) [inline apollo-federation/src/sources/connect/models/validation.rs:706]:
```suggestion
```

I don't want the main consumer to ever forget to translate new codes into their string representations.

File: apollo-federation/src/sources/connect/models/validation.rs [hunk 8]
```diff
@@ -311,12 +313,64 @@ fn validate_field(
...
+            connect_directive_url_coordinate(
+                connect_directive_name,
+                http_method,
+                object_name,
+                &field.name,
+            ),
+        ))
+    } else {
+        None
+    };
```
Comment (dylan-apollo) [inline apollo-federation/src/sources/connect/models/validation.rs:373]:
```suggestion
    let http_arg_url = http_methods.first().map(|(http_method, url)| {
        (
            url,
            connect_directive_url_coordinate(
                connect_directive_name,
                http_method,
                object_name,
                &field.name,
            ),
        )
    });
```

**Review by dylan-apollo — CHANGES_REQUESTED** [review]
Just a couple extra nits—love the updated snapshotting process!!

File: licenses.html (hunk 6 above)
Comment (pubmodmatt) [reply licenses.html:47]: That was my assumption, but I'll confirm.

File: apollo-federation/src/sources/connect/models/validation.rs (hunk 7 above)
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs:706]: Makes sense, I'll remove this.

File: apollo-federation/src/sources/connect/models/validation.rs (hunk 8 above)
Comment (pubmodmatt) [reply apollo-federation/src/sources/connect/models/validation.rs:373]: Nice, this evolved through a couple iterations and I somehow missed this simplification. I'll update it.

**Review by dylan-apollo — APPROVED** [review]
Looks great, thanks so much! Only thing I'm not positive about is the licenses file, but I'm assuming that's automated and CI would tell us if there was a problem.

File: licenses.html (hunk 6 above)
Comment (pubmodmatt) [reply licenses.html:47]: Confirmed that this is done via `cargo xtask all`, which in turn calls `cargo about` - see [licenses.rs](https://github.com/apollographql/router/blob/dev/xtask/src/commands/licenses.rs).

Comment (lennyburdette) [reply licenses.html:47]: I would remove it from the diff, that xtask isn’t deterministic so it’s just noise

Comment (pubmodmatt) [reply licenses.html:47]: If I were to do that, then every time someone runs `cargo xtask all` from now on, they'll get this diff locally... Seems better to just include it? But I'll defer to you @lennyburdette if you think it should be reverted.

Comment (dylan-apollo) [reply licenses.html:47]: Good to know, thanks! I wonder if this shouldn't be in `cargo xtask all`, then, and instead in a dedicated command 🤔. Maybe something we can ask the other contributors about.

Comment (pubmodmatt) [reply licenses.html]: I reverted the `licenses.html` changes

**Pull request merged**